iocage_lib/ioc_json.py

```diff
@@ -44,19 +44,672 @@
+                               f' Please destroy {uuid} and recreate'
+                               ' it.')
+
+        if release[:4].endswith('-'):
+            # 9.3-RELEASE and under don't actually have this binary.
+            release = conf['release']
+        elif release == 'EMPTY':
+            pass
+        else:
+            try:
+                with open(freebsd_version, 'r') as r:
+                    for line in r:
+                        if line.startswith('USERLAND_VERSION'):
+                            release = line.rstrip().partition('=')[2]
+                            release = release.strip("'\"")
+            except Exception as e:
+                iocage_lib.ioc_common.logit(
+                    {
+                        'level': 'EXCEPTION',
+                        'message': 'Exception:'
+                        f" '{e.__class__.__name__}:{str(e)}' occured\n"
+                        f"Loading {uuid}'s configuration failed"
+                    }
+                )
+
+            cloned_release = conf['release']
+
+        # Set all Version 3 keys
+        conf['release'] = release
+        conf['cloned_release'] = cloned_release
+
+        # Version 8 migration from uuid to tag named dataset
+        try:
+            tag = conf['tag']
+            uuid = conf['host_hostuuid']
+
+            try:
+                state = iocage_lib.ioc_common.checkoutput(
+                    ['jls', '-j', f'ioc-{uuid.replace(".", "_")}'],
+                    stderr=su.PIPE).split()[5]
+            except su.CalledProcessError:
+                state = False
+
+            if tag != uuid:
+                if not self.force:
+                    iocage_lib.ioc_common.logit(
+                        {
+                            'level': 'EXCEPTION',
+                            'message':
+                            'A renaming operation is required to continue.\n'
+                            f'Please run iocage -f {sys.argv[-1]} as root.'
+                        },
+                        _callback=self.callback,
+                        silent=self.silent)
+
+                conf, rtrn, date = self.json_migrate_uuid_to_tag(
+                    uuid, tag, state, conf)
+
+                conf['jail_zfs_dataset'] = f'iocage/jails/{tag}/data'
+
+                if not date:
+                    # The jail's tag was not a date, so it was renamed. Fix
+                    # fstab
+
+                    for line in fileinput.input(
+                            f'{self.iocroot}/jails/{tag}/fstab', inplace=1):
+                        print(line.replace(uuid, tag).rstrip())
+
+                    renamed = True
+
+                if rtrn:
+                    # They want to stop the jail, not attempt to migrate before
+
+                    return conf
+
+        except KeyError:
+            # New jail creation
+            pass
+
+        try:
+            if not renamed:
+                self.json_write(conf)
```

Comment:
Not sure if `json_write` can raise `FileNotFoundError` ? I think it will write to it if it doesn't exist